src/linalg.jl

```diff
@@ -339,28 +339,26 @@ function dot(A::AbstractSparseMatrixCSC{T1,S1},B::AbstractSparseMatrixCSC{T2,S2}
     return r
 end
 
-function dot(x::AbstractVector, A::AbstractSparseMatrixCSC, y::AbstractVector)
+function dot(x::AbstractVector{T1}, A::SparseMatrixCSC{T2}, y::AbstractVector{T3}) where {T1,T2,T3}
     require_one_based_indexing(x, y)
     m, n = size(A)
     (length(x) == m && n == length(y)) || throw(DimensionMismatch())
     if iszero(m) || iszero(n)
-        return dot(zero(eltype(x)), zero(eltype(A)), zero(eltype(y)))
+        return dot(zero(T1), zero(T2), zero(T3))
     end
-    T = promote_type(eltype(x), eltype(A), eltype(y))
-    r = zero(T)
-    rvals = getrowval(A)
-    nzvals = getnzval(A)
-    @inbounds for col in 1:n
+    T = promote_type(T1, T2, T3)
+    s = zero(T)
+    @inbounds @simd for col in 1:n
         ycol = y[col]
-        if _isnotzero(ycol)
-            temp = zero(T)
-            for k in nzrange(A, col)
-                temp += adjoint(x[rvals[k]]) * nzvals[k]
+        if !iszero(ycol)
+            for j in A.colptr[col]:(A.colptr[col+1]-1)
+                row = A.rowval[j]
+                val = A.nzval[j]
+                s += val * conj(x[row]) * ycol
```

Comment:
```suggestion
                s += dot(x[row], val, ycol)
```

`dot` acts recursively, so this was, strictly speaking, not correct even in the previous version.